File: Lab1/Cau2/Java/ProductSearchCriteria.java
import java.math.BigDecimal;
import java.util.Locale;

public final class ProductSearchCriteria {
    private final String keyword;
    private final String category;
    private final BigDecimal minPrice;
    private final BigDecimal maxPrice;
    private final int page;
    private final int pageSize;

    private ProductSearchCriteria(Builder builder) {
        keyword = normalize(builder.keyword);
        category = normalize(builder.category);
        minPrice = builder.minPrice;
        maxPrice = builder.maxPrice;
        page = builder.page;
        pageSize = builder.pageSize;
    }

    private static String normalize(String value) {
        if (value == null) return null;
        String normalized = value.replaceAll("(?U)\\s+", " ").strip().toLowerCase(Locale.ROOT);
        return normalized.isEmpty() ? null : normalized;
    }

    public String getKeyword() { return keyword; }
    public String getCategory() { return category; }
    public BigDecimal getMinPrice() { return minPrice; }
    public BigDecimal getMaxPrice() { return maxPrice; }
    public int getPage() { return page; }
    public int getPageSize() { return pageSize; }

    public static final class Builder {
        private String keyword;
        private String category;
        private BigDecimal minPrice;
        private BigDecimal maxPrice;
        private int page = 1;
        private int pageSize = 5;

        public Builder keyword(String value) { keyword = value; return this; }
        public Builder category(String value) { category = value; return this; }
        public Builder minPrice(BigDecimal value) { minPrice = value; return this; }
        public Builder maxPrice(BigDecimal value) { maxPrice = value; return this; }
        public Builder page(int value) { page = value; return this; }
        public Builder pageSize(int value) { pageSize = value; return this; }

        public ProductSearchCriteria build() {
            if (page < 1)
                throw new IllegalArgumentException("Page must be at least 1.");
            if (pageSize < 1 || pageSize > 100)
                throw new IllegalArgumentException("PageSize must be between 1 and 100.");
            if ((minPrice != null && minPrice.signum() < 0) ||
                    (maxPrice != null && maxPrice.signum() < 0))
                throw new IllegalArgumentException("Prices must be non-negative.");
            if (minPrice != null && maxPrice != null && minPrice.compareTo(maxPrice) > 0)
                throw new IllegalArgumentException("MinPrice must not exceed MaxPrice.");
            return new ProductSearchCriteria(this);
        }
    }
}

File: Cau2/CSharp/ProductSearchCriteria.cs
using System.Text.RegularExpressions;

namespace Cau2;

public sealed class ProductSearchCriteria
{
    public string? Keyword { get; }
    public string? Category { get; }
    public decimal? MinPrice { get; }
    public decimal? MaxPrice { get; }
    public int Page { get; }
    public int PageSize { get; }

    private ProductSearchCriteria(Builder builder)
    {
        Keyword = Normalize(builder.KeywordValue);
        Category = Normalize(builder.CategoryValue);
        MinPrice = builder.MinPriceValue;
        MaxPrice = builder.MaxPriceValue;
        Page = builder.PageValue;
        PageSize = builder.PageSizeValue;
    }

    private static string? Normalize(string? value)
    {
        if (value is null) return null;
        var normalized = Regex.Replace(value, @"\s+", " ").Trim().ToLowerInvariant();
        return normalized.Length == 0 ? null : normalized;
    }

    public sealed class Builder
    {
        internal string? KeywordValue { get; private set; }
        internal string? CategoryValue { get; private set; }
        internal decimal? MinPriceValue { get; private set; }
        internal decimal? MaxPriceValue { get; private set; }
        internal int PageValue { get; private set; } = 1;
        internal int PageSizeValue { get; private set; } = 5;

        public Builder Keyword(string? value) { KeywordValue = value; return this; }
        public Builder Category(string? value) { CategoryValue = value; return this; }
        public Builder MinPrice(decimal? value) { MinPriceValue = value; return this; }
        public Builder MaxPrice(decimal? value) { MaxPriceValue = value; return this; }
        public Builder Page(int value) { PageValue = value; return this; }
        public Builder PageSize(int value) { PageSizeValue = value; return this; }

        public ProductSearchCriteria Build()
        {
            if (PageValue < 1)
                throw new ArgumentOutOfRangeException(nameof(Page), "Page must be at least 1.");
            if (PageSizeValue is < 1 or > 100)
                throw new ArgumentOutOfRangeException(nameof(PageSize), "PageSize must be between 1 and 100.");
            if (MinPriceValue < 0 || MaxPriceValue < 0)
                throw new ArgumentException("Prices must be non-negative.");
            if (MinPriceValue > MaxPriceValue)
                throw new ArgumentException("MinPrice must not exceed MaxPrice.");
            return new ProductSearchCriteria(this);
        }
    }
}
